pkg/netceptor/conn_internal_test.go

```diff
@@ -0,0 +1,870 @@
+			// Verify the result content
+			if result.Conn != testConn {
+				t.Errorf("Expected conn %v, got %v", testConn, result.Conn)
+			}
+			if result.Err != nil {
+				t.Errorf("Expected nil error, got %v", result.Err)
+			}
+		case <-time.After(100 * time.Millisecond):
+			t.Error("Expected result to be sent, but it wasn't")
+		}
+	})
+}
+
+// mockNetceptorForListen is a mock for Netceptor used in listen tests.
+type mockNetceptorForListen struct {
+	listenerRegistry  map[string]*PacketConn
+	reservedServices  map[string]func(*MessageData) error
+	nodeID            string
+	maxForwardingHops byte
+	logger            *logger.ReceptorLogger
+	context           context.Context
+	cancelFunc        context.CancelFunc
+	listenerLock      sync.RWMutex
+	getEphemeralFunc  func() string
+}
+
+// listen is a simplified implementation of Netceptor.listen for testing.
+func (m *mockNetceptorForListen) listen(ctx context.Context, service string, tlscfg *tls.Config, advertise bool, adTags map[string]string) (*Listener, error) {
+	if len(service) > 8 {
+		return nil, fmt.Errorf("service name %s too long", service)
+	}
+	if service == "" {
+		service = m.GetEphemeralService()
+	}
+	m.listenerLock.Lock()
+	defer m.listenerLock.Unlock()
+	_, isReserved := m.reservedServices[service]
+	_, isListening := m.listenerRegistry[service]
+	if isReserved || isListening {
+		return nil, fmt.Errorf("service %s is already listening", service)
+	}
+
+	// In a real implementation, we would create a PacketConn, set up QUIC, etc.
+	// For testing, we'll just return a minimal Listener
+
+	if advertise {
+		m.AddLocalServiceAdvertisement(service, 0, adTags)
+	}
+
+	doneChan := make(chan struct{})
+	li := &Listener{
+		s:          nil, // We don't need this for the test
+		pc:         nil, // We don't need this for the test
+		ql:         nil, // We don't need this for the test
+		AcceptChan: make(chan *AcceptResult),
+		DoneChan:   doneChan,
+		doneOnce:   &sync.Once{},
+	}
+
+	return li, nil
+}
+
+func newMockNetceptorForListen() *mockNetceptorForListen {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	return &mockNetceptorForListen{
+		listenerRegistry:  make(map[string]*PacketConn),
+		reservedServices:  make(map[string]func(*MessageData) error),
+		nodeID:            "test-node",
+		maxForwardingHops: 5,
+		logger:            logger.NewReceptorLogger("test"),
+		context:           ctx,
+		cancelFunc:        cancel,
+		listenerLock:      sync.RWMutex{},
+		getEphemeralFunc:  func() string { return "ephemeral-service" },
+	}
+}
+
+func (m *mockNetceptorForListen) GetEphemeralService() string {
+	return m.getEphemeralFunc()
+}
+
+func (m *mockNetceptorForListen) AddNameHash(name string) uint64 {
+	return 0
+}
+
+func (m *mockNetceptorForListen) AddLocalServiceAdvertisement(service string, connType byte, tags map[string]string) {
+	// Mock implementation - do nothing
+}
+
+func (m *mockNetceptorForListen) GetLogger() *logger.ReceptorLogger {
+	return m.logger
+}
```

Comment:
The pattern for these mocks is to generate them into  `mock_netceptor` rather than writing them in-line in the test file.